Assignment 1
==============

# Template for Unix, git, and Sequence Alignment Assignments (20 pts possible)

__Full Name:__

__Student ID:__

*_Any code should be formatted as such_*

## Markdown

Paste the bio of your the student you interviewed here:


## Unix

Paste the `date` command that you used for exercise one here:

## Git

Paste the URL for the shared github repository that you created with your lab partners here.  Use proper markdown formatting:




## Sequence Alignment

*_Please answer the following questions Be clear and concise with any written answers._*

__1.__ Paste in the markdown table from the lab manual that includes for each genome:

* Size of the file
* Number of chromosomes
* Size of the genome in bp
* Number of protein-coding genes
* Average protein length

For _ONE_ of the files, provide the code that you used to answer these questions.

__2.__ How do you know that when you use `shuffleseq` the sequences have the same exact composition?

__3.__ Create a text based "histogram" as described in the lab manual
that shows the distribution of scores when aligning shuffled sequences.
Why is the shape of the curve not quite normal?

__4.__ What would be the effect of doing 1,000,000 shuffled sequences?

__5.__ Briefly discuss the relationship of sequence length and score.

__6.__ Starting with the C. elegans B0213.10 protein, find the best
match in the A. thaliana and D. melanogaster proteomes with `water`.
Record their alignment scores, percent identities, and protein names
here.  (Use a markdown table!)

__7.__ What is the expected (average) score of each pairwise alignment
at random? (Perform some shuffled alignments to get an idea of what the
random expectation is).

__8.__ How many Z-scores away from the mean is the best alignment?

__9.__ Briefly discuss the statistical and biological significance of your results.

__10.__ Approximately how long would it take to compare two proteomes using `water`?
