---
layout: lab
hidden: true
title: 'Differential Gene Expresion: Now What?'
output: html_document
tags:
- R
- Brassica
- Illumina
- RNAseq
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

# Assignment

The three 

# Background

Now what?

We have found our lists of differentially expressed genes.  Now what?  We want to know the potential functions of those differentially expressed genes and possibly their regulators.  Specifically we will:

1. Merge the listd of differentially expressed genes with a gene annotation file to learn about the function of the most upregulated genes.
2. Ask if particular classes of genes genes are upregualted, by using functional classifiers known as [Gene Ontology (GO) terms](http://geneontology.org/).
3. Determine whether the differentially expressed genes have any common promoter motifs.
4. Group genes based on their pattern of gene expression across the samples.

## Annotate the differentially expressed genes.

As part of [his B_rapa annotation paper](www.g3journal.org/content/4/11/2065.long) a postdoc in my lab, Upendra Devisetty, did a quick annotation of the _B. rapa_ genes by BLASTing each gene to the NCBI non-redundant database and taking the gene description of the best match (above a threshold). You can download the description with this command:

  wget http://www.g3journal.org/content/suppl/2014/08/12/g3.114.012526.DC1/FileS9.txt
  
Place the file into your `Brapa_reference` directory.
  
Now open Rstudio, set the working directory to your `Diff_Exp` directory, and proceed.
  
__Exercise 1__

__a.__ Use `merge()` to add gene descriptions for the genes found to be regualted by the DP treatment.  Output a table of the top 10 genes that includes the output from edgeR and the descriptions.  __Important: Pay attention to the "sort="" argument to `merge()`.  Should it be TRUE or FALSE?

__b.__ Repeat this for  genes with a genotype x trt interaction.

```{r Ex1 ,echo=FALSE,eval=FALSE}
#a
annotation <- read.delim("../data/FileS9.txt",header=FALSE,row.names=1) #path will be different for students
head(annotation)
DEgene.trt <- read.csv("DEgenes.trt.csv",row.names=1)
head(DEgene.trt)
DE.trt.annotated <- merge(DEgene.trt,annotation,by="row.names",sort=FALSE)
head(DE.trt.annotated,10)

#b
DEgene.interaction <- read.csv("DEgenes.interaction.csv",row.names=1)
head(DEgene.interaction)
DE.interaction.annotated <- merge(DEgene.interaction,annotation,by="row.names",sort=FALSE)
head(DE.interaction.annotated,10)
```

By looking at the annotated interaction gene list we can see that many of the identified genes code for proteins that modify the plant cell wall (I wouldn't expect you to know this unless you are a plant biologist).  This might relate to the different properties of the two cultivars, IMB211 and R500, and their different responses to the treatment.  Depending on our interests and knowledge of the sytsem we might at this point choose follow up study on specific genes.

## Test for enrichment of functional classes of genes.

A casual glance indicated that there might be an enrichment for cell wall related genes in  the gt X trt DE gene list.  We can test this more rigorously by asking if there is statistical over-representation of particular [Gene Ontology (GO)](http://geneontology.org/) terms in this set of genes.  GO terms provide a precise, defined language to describe gene function.

To test for test for enrichment of GO terms we essentially ask the question of whether the prevalence of a term in our gene set of interest is higher than its prevalence in the rest of the genome (aka the universe).  For example if 20% of the differentially expressed genes have the GO term "Cell Wall" but only 10% of the not-differentially expressed genes have the term Cell Wall that the term "Cell Wall" might be over-represented, or enriched, in our differentially expressed genes.  In principle this could be tested using a [Chi-squared test](http://www.biostathandbook.com/chiind.html) or [Fisher's exact test](http://www.biostathandbook.com/fishers.html) for contigency tables.  In practice we will use [GOseq](http://www.bioconductor.org/packages/release/bioc/html/goseq.html) that [makes an adjustement for gene-length bias](http://genomebiology.com/2010/11/2/r14) since it is easier to detect differential expression of longer genes.  

It is important in these analyses to define the "Universe" correctly.  The Universe of genes is all of the genes where you could have detected an expression difference.  It would not have been possible to detect an expression difference in genes that were not expressed at a detectable level in our expereiment.  So the Universe in this case is all expressed genes, rather than all genes.

### Install GOseq

```{r, eval = FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("goseq")
```

### Download the GO annotation and gene length for B.rapa

GO term annotation for _B. rapa_ is also available from the [Devisetty et al paper](http://www.g3journal.org/content/4/11/2065.long).

cDNA gene lengths were estimated by the `featureCounts()` command used in Tuesday's lab.  I have saved them for you and you can download them as instructed below.  (You're welcome)

Download the files usind the commands below (in Linux) and place the file in your `Brapa_reference` directory.

GO terms:

    wget http://www.g3journal.org/content/suppl/2014/08/12/g3.114.012526.DC1/FileS11.txt
    
Gene Length:

    wget {{site.baseurl}}/data/Brapa_CDS_lengths.txt
  
### Format data for GOseq

We need to do a bit of data wrangling to get things in the correct format for GOSeq.  First we import the gene lengths and GO terms.  We also import a table of all expressed genes in the experiment (you could have gotten this from Tueday's data)

Get the data
```{r goseq_prep, eval=FALSE}
library(goseq)
go.terms <- read.delim("../Brapa_reference/FileS11.txt",header=FALSE,as.is=TRUE)
head(go.terms)
names(go.terms) <- c("GeneID","GO")
summary(go.terms)

expressed.genes <- read.delim("internode_expressed_genes.txt",as.is=TRUE)
head(expressed.genes)
names(expressed.genes) <- "GeneID"

gene.lengths <- read.table("../Brapa_reference/Brapa_CDS_lengths.txt",as.is=TRUE)
head(gene.lengths)
summary(gene.lengths)

#we need to reduce the gene.length to only contain entries for those genes in our expressed.genes set.  We also need this as a vector
gene.lengths.vector <- gene.lengths$Length[gene.lengths$GeneID %in% expressed.genes$GeneID]
names(gene.lengths.vector) <- gene.lengths$GeneID[gene.lengths$GeneID %in% expressed.genes$GeneID]
head(gene.lengths.vector)

#Do the reverse to make sure everyting matches up (it seems that we don't have length info for some genes?)
expressed.genes <- expressed.genes[expressed.genes$GeneID %in% names(gene.lengths.vector),]
```

Format go.terms for goseq.  We want them in list format, and we need to separate the terms into separate elemants.
```{r goseq_prep2,eval=FALSE}
go.list <- strsplit(go.terms$GO,split=",")
names(go.list) <- go.terms$GeneID
head(go.list)
```

Format gene expression data for goseq.  We need a vector for each gene with 1 indicating differential expression and 0 indicating no differential expression.
```{r goseq_prep3, eval=FALSE}
DE.interaction <- expressed.genes %in% rownames(DEgene.interaction) 
    #for each gene in expressed gene, return FALSE if it is not in DEgene.trt and TRUE if it is.
names(DE.interaction) <- expressed.genes$GeneID
head(DE.interaction)
DE.trt <- as.numeric(DE.interaction) #convert to 0s and 1s
head(DE.interaction)
sum(DE.interaction) # number of DE genes
```

### Calculate over-representation

Now we can look for GO enrichment
```{r goseq}
#determines if there is bias due to gene length.  The plot shows the relationship.
nullp.result <- nullp(DEgenes = DE.interaction,bias.data = gene.lengths.vector)

#calculate p-values for each GO term
rownames(nullp.result) <- names(gene.lengths.vector) #because of a bug in nullp()
GO.out <- goseq(pwf = nullp.result,gene2cat = go.list,test.cats=("GO:BP"))

#list over-represented GO terms (p < 0.05)
GO.out[GO.out$over_represented_pvalue < 0.05,]
```

### GO visualization

Looking through a long list can be tough.  There is a nice visualizer called [REVIGO](http://revigo.irb.hr/).  To use it we need to cut and paste the column with the GO term and the one with the p-value.  Use the command below to print these to columns to the console:
```{r goseq_revigo}
print(GO.out[GO.out$over_represented_pvalue < 0.05,1:2],row.names=FALSE)
```

Cut and paste the GO terms and p-values into [REVIGO](http://revigo.irb.hr/).  You can use the default settings.  There are three types of GOterms:

* Biological Process (BP)
* Cellular Compartment (CC)
* Molecular Functions (MF)

Generally I find the BP terms most helpful but you can look at each type by clicking in the tabs.

REVIGO has three types of visualizers.  The "TreeMap"  Can be particular nice because it groups the GO terms into a hierarchy.

__Exercise 2__:  

__a:__ In REVIGO display a "TreeMap" of the BP GO terms.  Was our hypothesis that cell wall genes are enriched in the genotype X treatment gene set correct?  You DO NOT need to include the treemap in your answer.

__b:__ Display a "TreeMap" of the CC GO terms.  There are four general categories shown, some with sub-categories.  What are the two general categories with the largest number of sub categories?  How might these general categories relate to differences in plant growth? 

