---
layout: lab
title: Genetic Networks 2 - Graphs
hidden: true
tags:
     - networks
     - R
     - clustering
     - kmeans
     - RNA-seq 
     - graphs
--- 
# Biological Context Review
In the past few weeks we have discussed taking data from off the sequencer, aligned the reads to our reference genome, calculated counts for the number of reads that mapped to each gene in out reference genome, found out what genes are differentially expressed between our genotype treatment combinations, and started to interpret the data using clustering. With these large data sets there is more than one way to look at the data. As biologists this is where we need to critically evaluate what the data is telling us and interpret it using our knowledge of biology and the design of the experiment. In our case we have an environmental perturbation that we have imposed on the plants by altering the density of plants in a given pot. Understanding mechanistically how plants respond to crowding is important for understanding how plants grow and compete for resources in natural ecosystems, and how we might manipulate plants to grow optimally in agroecosystems. In our case, we have two genotypes of plants that show very different physiological and morphological responses to crowding. We have a lot of data, some quantitative, some observational, that support this. Plant growth, just like the growth of any organism, is a really complicated thing. Organisms have evolved to interact with the environment by taking in information from their surroundings and trying to alter thier physiology and biochemistry to better live in that environment. We know a lot about the details of how these signals are intercepted by the organisms, but we know less how this translates to changes in biochemistry, physiology, development and ultimately growth. In our case, plants recieve information about thier neighbors through detecting changes in light quality through the phytochrome light receptors. This is a focus of the Maloof lab. You can read more generally about the way plants percieve these changes **here**. We want to understand how plants connect the upstream perception of environmental signals (in this case the precense of neighbors) and how this information cascades through the biological network of the organism to affect the downstream outputs of physiological and developmental changes. To get an approximation of what is going on in the biological network we need to work with an intermediate form of biological information: gene expression. Although there are important limitations to only using only gene expression data which we will discuss during the lecture, it should provide some clues as to how to connect the upstream environmental perception with the downstream outputs.

















# Biological Networks
# Clustering Introduction
As you learned last week, when we are dealing with genome scale data it is hard to come up with very specific summaries of the data unless you know exactly the question you are trying to ask computationally. Today we will be talking about three different ways to cluster data and get visual summaries of the expression of all genes that had a significant GxE interaction. Once we have these clusters, it allows us to ask further, more detailed questions such as what GO categories are enriched in each cluster, or are there specfic metabolic pathways contained in the clusters? While clustering can be used in an exploratory way, the basics you will be learning today have been extended to very sophisticated statistical/machine learning methods used across many disciplines. In fact, there are many different methods used for clustering in R outlined in this **[CRAN VIEW](http://cran.r-project.org/web/views/Cluster.html)**.

# Hierarchical Clustering
An intuitive example is clustering the distances between know geographic locations, such as US Cities. I took this example from this [website](http://www.analytictech.com/networks/hiclus.htm).
 
1. Calculate a distance measure between all the row and column combinations in data set (think geographic distances between cities)
2. Start all the items in the data out as belonging to its own cluster (every city is its own cluster)


```{r, engine = 'bash', eval = FALSE}
wget https://github.com/rjcmarkelz/BIS180L_web/tree/feature-networks/data/us_cities.txt
```

```{r  warning=FALSE} 
# make sure to change the path to where you downloaded this using wge
```
**Exercise 1:**
Extending the example that I gave for BOS/NY/DC, what are the distances that define each split in the Westcoast side of the hclust plot? 
*Hint 1: Start with the distances between SF and LA. Then look at the difference between that cluster up to SEA*
*Hint 2: Print cities, you only need to look at the upper right triangle of data matrix.*













