---
type: lab
hidden: true
title: Differential Gene Expression from Illumina RNAseq
output: html_document
---

## Background

## Preliminaries

We need a couple of additional Bioconductor libraries.  The `Rsubread` library will enable us to go from mapped reads to a count of reads per gene.  The `rtracklayer` library enables us to convert gff files to gtf files.

First we need to install a Linux dependency.  From the Linux command line:

  sudo apt-get install libxml2-dev

Now we can install the R pacakages.  Start Rstudio and enter:

```{r install_subread,eval=FALSE}
  source("http://bioconductor.org/biocLite.R")
	biocLite(c("Rsubread","rtracklayer"))
```

If R asks you about using a personal library, answer "y".  If R asks you if you want to update packages you can answer "n". 

Don't worry about the warning messages.  Do worry if you get a message saying "ERROR could not install."

## GFF to GTF

There are two closely related file formats that describe genomic features, [GFF](http://www.sequenceontology.org/gff3.shtml) and [GTF](http://mblab.wustl.edu/GTF22.html).  Unfortunately we have a GFF but we need a GTF.  Fortunately the conversion is pretty easy.

Set your working directory to the Brassica_Assignment directory.  Then:

```{r GFF2GTF}
library(rtracklayer)
gff <- import.gff("Brapa_reference/Brapa_gene_v1.5.gff")
gff #take a look

#create a column "gene_id" that contains the gene name for every entry
gff$gene_id <- regmatches(gff$group,regexpr("Bra[0-9]+$",gff$group))

export(gff,"Brapa_reference/Brapa_gene_v1.5.gtf",format="gtf")
```

## Bam to read counts

As you know from last week's lab, we mapped RNAseq reads to the _B. rapa_ genome.  In order to ask if reads are differentially expressed between cultivars (IMB211 vs. R500) or treatments (dense planting vs. non-dense planting) we need to know how many reads were sequences from each gene.

To do this we use the bam files (telling us where in the genome the reads mapped) and the .gtf file that we just generated (telling us where the genes are) to figure out which reads belong to which genes.  Thankfully the `Rsubread` package does this for us.  An alternative workflow (not used here) would be to use the python package [`HTSeq`](http://www-huber.embl.de/HTSeq/)

Let's try this on the two files from Thursday.  You may need to change the path listed below to make this work.

```{r Rsubread}
library(Rsubread)
readCounts <- featureCounts(
  files=c("tophat_out-IMB211_All_A01_INTERNODE.fq/accepted_hits_A01.bam","tophat_out-R500_All_A01_INTERNODE.fq/accepted_hits_A01.bam"),
  annot.ext="Brapa_reference/Brapa_gene_v1.5.gtf", 
  isGTFAnnotationFile=TRUE,
  GTF.featureType="CDS", # This depends on GTF file.  Often it would be "exon"
  GTF.attrType="gene_id"
  )
```

__Exercise 1__  
Read the help file for feature counts.  Be sure to look at the section "Value" where it describes the output.  
__a__ Provide a line of code that displays the counts of the first 6 genes.  (It is not very interesting because the first genes in the file are on chromosome A03 (strange numbering...) and our bam file only has counts from A01...  )
__b__ The gene `Bra011030` is on chromosome A01.  What are its read counts in the two files?
__c__ What percentage of reads (from each file) were assigned to a gene?  What percentage were unassigned because they were not located in a gene (aka "Feature")?
__d__ What are 2 possible reasons why there are reads that cannot be assigned to a gene?  
```{r Ex1,echo=FALSE,eval=FALSE}
head(readCounts$count)
readCounts$count["Bra011030",]
readCounts$stat
100*readCounts$stat[,-1]/colSums(readCounts$stat[,-1])
```

