---
title: "Illumina continued: analyze VCF file in R"
layout: lab
tags:
- R
- Brassica
- Illumina
hidden: true
output: html_document
---

In the first part of today's lab you used `FreeBayes` to calculate SNPs between IMB211 and the reference sequence and betwen R500 and the reference.  `FreeBayes` returned a [.vcf (variant call format)](http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40_) file.
    
The .vcf file contains information on polymorphisms.  It is a tab-delimited file that is easily imported into R.  We want to filter the SNPs to remove low quality SNPs.  We also could use this for downstream analyses, i.e. use these as genetic markers for mapping or QTL mapping, look for genes with coding changes, etc., although we will not have time to explore those types of analyses in this class.

```{r getdata, eval=FALSE}
library(ggplot2)
vcf.data <- read.table("IMB211_R500.vcf",as.is=T,na.strings = ".")
head(vcf.data)
```

```{r getdata_jm, echo=FALSE, eval=FALSE}
library(ggplot2)
vcf.data <- read.table("../data/IMB211_R500.vcf",as.is=T,na.strings = ".")
head(vcf.data)
```

There is no column heading in what we imported.  There is a column heading in the .vcf file but is not loaded into R because its first character is "#".  Here we get it by using a `system()` call.  `system()` causes R to issue a command to the Linux shell.  The column headings in the .vcf start with "#CHROM" so we use Linux's version of `grep` to search for that line.  The argument `intern = TRUE` tells R that we want to capture the output of the command.

```{r getheader,eval=FALSE}
vcf.header <- system("grep '#C' IMB211_R500.vcf",intern = TRUE) #might not work on Windows
vcf.header
vcf.header <- sub("#","",vcf.header) #get rid of the pound sign
```

```{r getheader_jm,echo=FALSE,eval=FALSE}
vcf.header <- system("grep '#C' ../data/IMB211_R500.vcf",intern = TRUE) #might not work on Windows
vcf.header
vcf.header <- sub("#","",vcf.header) #get rid of the pound sign
```

The header is tab delimited, so we split on tabs and use those as column names
```{r formatheader,eval=FALSE}
vcf.header <- unlist(strsplit(vcf.header,split="\t"))
colnames(vcf.data) <- vcf.header
head(vcf.data)
```

Now with the headers attached the file makes a bit more sense.  Detailed information [is here](https://samtools.github.io/hts-specs/VCFv4.2.pdf), but briefly:

* CHROM and POS should be obvious
* ID is not used here
* REF is the reference sequence (From the fasta file)
* ALT is the allele found in our sample(s)
* QUAL is the PHRED scaled probability that the locus is actually homozygous for the reference allele, or 1 - P(locus is homozygous for the alternate allele).  So QUAL of 20 means that there is a 0.01 probablity that the locus is actually not polymorphic. _Note that the definition of this field varies depending on which SNP calling program you use._
* FILTER is not used here.  There are programs that will place flags in this field to indicate that a SNP should be filtered out.
* INFO lots of INFO.  We can learn about these from the beginning of the vcf file (See below)
* FORMAT tells is the identity of the numbers in the next two fields
    * GT: The most probable genotype.  1/1 is homozygous alternate; 0/0 is homozygous reference; 0/1 is heterozygous.
    * DP: Read Depth
    * DPR: Number of observation for each allele.  (Actually I think the first number is total counts and the second is alt counts)
    * RO: Reference allele observation count
    * QR: Sum of quality of the reference observations
    * AO: Alternate allele observation count
    * QA: Sum of quality of the alternate observations
    * GL: Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy

This information is at the beginning of the .vcf file which we could look at with `less` at the Linux command line, or we can look at it in R using system calls.  (Might not work on Windows)

```{r moreHeader,eval=FALSE}
system("grep '##INFO' IMB211_R500.vcf")

system("grep '##FORMAT' IMB211_R500.vcf")
```

```{r moreHeader_jm,echo=FALSE,eval=FALSE}
system("grep '##INFO' ../data/IMB211_R500.vcf",intern=TRUE)

system("grep '##FORMAT' ../data/IMB211_R500.vcf",inter=TRUE)
```

To be able to use the SNP data in R we need to break up the data in the final two columns.  We use the command `strsplit()` for this.  `strsplit()` splits a character string based at every occurrence of a character.  We will make a temporary matrix to hold the split data and then bind it back to the original data file.

```{r split,eval=FALSE}
# Before splitting add NAs to blank cells

vcf.data$IMB211[is.na(vcf.data$IMB211)] <- "NA:NA:NA:NA:NA:NA:NA:NA"

IMB211.tmp <- matrix(
  unlist(strsplit(vcf.data$IMB211,split = ":")),
  nrow=nrow(vcf.data),
  byrow=TRUE
  )
head(IMB211.tmp)
colnames(IMB211.tmp) <- paste("IMB211",c("gt","tot.depth","allele.obs","ref.depth","ref.qual","alt.depth","alt.qual","gen.lik"),sep="_")

vcf.data$R500[is.na(vcf.data$R500)] <- "NA:NA:NA:NA:NA:NA:NA:NA"

R500.tmp <- matrix(
  unlist(strsplit(vcf.data$R500,split = ":")),
  nrow=nrow(vcf.data),
  byrow=TRUE
  )
head(R500.tmp)
colnames(R500.tmp) <- paste("R500",c("gt","tot.depth","allele.obs","ref.depth","ref.qual","alt.depth","alt.qual","gen.lik"),sep="_")

vcf.data <- cbind(vcf.data,IMB211.tmp,R500.tmp,stringsAsFactors=FALSE)
summary(vcf.data)
```

There is one problem: we need to convert the columns that should be numbers back to numeric.  This is done by extracting the relevant columns, using `apply` to apply the `as.numeric` function to each column, and then placing the results back in the correct place

```{r numeric,eval=FALSE}
head(vcf.data)
vcf.data[,c("IMB211_tot.depth","IMB211_ref.depth","IMB211_ref.qual","IMB211_alt.depth","IMB211_alt.qual",
            "R500_tot.depth","R500_ref.depth","R500_ref.qual","R500_alt.depth","R500_alt.qual")] <- 
  apply(vcf.data[,c("IMB211_tot.depth","IMB211_ref.depth","IMB211_ref.qual","IMB211_alt.depth","IMB211_alt.qual",
            "R500_tot.depth","R500_ref.depth","R500_ref.qual","R500_alt.depth","R500_alt.qual")],
        2,
        as.numeric
        )
head(vcf.data)
summary(vcf.data)
```

__Exercise 1__  
To explore the quality of our data, make a histogram of genotype quality.  It is hard to get a feel for the QUAL scores at the low end, so try making a second histogram that illustrates this region better. (Hint: one option is to subset the data)
```{r histograms, echo=FALSE, eval=FALSE}
hist(vcf.data$QUAL)
hist(vcf.data$QUAL[vcf.data$QUAL<500],breaks = 50)
```

__Exercise 2__  
We only want to keep positions that have a reasonable probabilty of being correct.  
__a__ At a quality of 40 what is the probability that the genotype call is wrong?
__b__Subset the data to only keep positions where the quality score is 40 or greater.  
__c__What percentage of SNPs were retained?

```{r filter, echo=FALSE,eval=FALSE}
vcf.data.good <- vcf.data[vcf.data$QUAL>=40,]
```

We can count the number of homozygous and heterozygous snps using table:

```{r snptable,eval=FALSE}
table(vcf.data.good$IMB211_gt)
table(vcf.data.good$R500_gt)
```

We can even count the numbers common and unique to each genotype

```{r eval=FALSE}
ftable(vcf.data.good[,c("IMB211_gt","R500_gt")])
```

__Exercise 3__  
__a__ Which SNPS would be most useful for a downstream QTL analysis between IMB211 and R500?  (Ignore the allele categories that have "2", "3", or "4")  
__b__ (optional) Subset the data frame so that you only have these SNPs.  
__c__ (optional) How many SNPS are retained?

__Exercise 4__  
__a__ Using the higher quality SNP list, plot the position along the chromosome of each SNP (x axis), read depth (R500 and IMB211 combined) (y axis).  
__b__ Why might there be gaps with no SNPs?  
__Optional__: color each SNP for whether it is R500 specific, IMB specific, or common.
```{r, eval=FALSE, echo=FALSE}
library(ggplot2)
vcf.data.good$category <- NA                             
vcf.data.good$category[vcf.data.good$IMB211_gt=="0/0"]  <- "R500_only"
vcf.data.good$category[vcf.data.good$R500_gt=="0/0"] <- "IMB211_only"
vcf.data.good$category[vcf.data.good$R500_gt!="0/0" & vcf.data.good$IMB211_gt!="0/0"] <- "Both"
table(vcf.data.good$category)
vcf.data.good$tot.depth <- vcf.data.good$IMB211_tot.depth+vcf.data.good$R500_tot.depth
qplot(x=POS,y=tot.depth,color=category,pch=category,data=vcf.data.good)
```

__For Fun (?)--not graded--__  
Plot the number of each type of base change (A->G, etc).  Are there differences?  Is this expected?
