---
title: "BLAST part 2"
author: "Julin N Maloof"
date: "4/19/2018"
output: html_document
---

For this lab we will continue to expand our R toolkit and apply what we learn to the BLAST results.

## Joins

Often we will have different data sets that we want to combine in some way.  For example, we might want to combine our reciprocal BLAST results into a single data frame.

### Join Tutorial

To learn how to do this please complete the tutorial that I wrote on joins.  You will need to re-install the tutorial package and then start the tutorial:

```{r eval=FALSE}
devtools::install_github("UCDBIS180L/BIS180LTutorials") # needs to be reinstalled to get the new tutorial

learnr::run_tutorial("Joins", package = "BIS180LTutorials") 
```

### Join the blast results

Let's say we want to find *all* of the Arabidopsis genes with reciprocal best hits in the worm genome.  How would we do that?  Let's think of the steps:

1. For each query record the E-value of the second-best hit.
2. Filter our hits to retain only the best hit. 
3. Join the two data sets together
4. Filter to keep genes with reciprical best hits.

Load the data.  Adjust the paths to work on your computer.
```{r}
library(tidyverse)
plant_worm <- read_tsv("../data/plant_vs_worm.blastout.gz",col_names=c("query_id",
                                                                       "subject_id",
                                                                       "pct_ident",
                                                                       "len",
                                                                       "mis",
                                                                       "gaps",
                                                                       "qb",
                                                                       "qe",
                                                                       "sb",
                                                                       "se",
                                                                       "E",
                                                                       "Score"))
worm_plant <- read_tsv("../data/worm_vs_plant.blastout.gz",col_names=c("query_id",
                                                                       "subject_id",
                                                                       "pct_ident",
                                                                       "len",
                                                                       "mis",
                                                                       "gaps",
                                                                       "qb",
                                                                       "qe",
                                                                       "sb",
                                                                       "se",
                                                                       "E",
                                                                       "Score"))

```

Make a column recording the next-best E-Score

```{r}
plant_worm <- plant_worm %>% 
  arrange(query_id, E) %>% 
  group_by(query_id) %>% 
  mutate(nextE = lead(E))
head(plant_worm,12)

worm_plant <- worm_plant %>% 
  arrange(query_id, E) %>% 
  group_by(query_id) %>% 
  mutate(nextE = lead(E))
head(worm_plant,12)
```

** Exercise 1 **
Look at the help page for `lead` and examine the `E` and `nextE` columns.  What does `lead()` do?  Why are there NAs in the `nextE` column?

Now lets filter these to keep the best hit, as we did in the last lab.  Also use select to only keep relevant columns.

```{r}
plant_worm_best <- plant_worm %>%
    arrange(query_id, E, desc(Score)) %>%
    filter(!duplicated(query_id)) %>%
    select(query_id, subject_id, pct_ident, len, E, nextE, Score)
```

```{r}
worm_plant_best <- worm_plant %>%
    arrange(query_id, E, desc(Score)) %>%
    filter(!duplicated(query_id)) %>%
    select(query_id, subject_id, pct_ident, len, E, nextE, Score)
```


The `subject_id` in the plant_worm_best set is too long and needs to be shortened to just the gene_id.  `str_replace` searches for its second argument and replaces it with its third.  We are using wildcards (techinically regular expressions) to search for strings that begin with "#" and replace them with nothing.  Hopefully there will be time to explain this in more detail in a later lab.

```{r}
head(plant_worm_best$subject_id) #yikes!

plant_worm_best <- plant_worm_best %>%
  mutate(subject_id=stringr::str_replace(subject_id,"#.*",""))

head(plant_worm_best$subject_id) #better

```

Now let's create a table of Arabidopsis genes that includes the reciprocal blast results

```{r}
plant_with_recip <- left_join(plant_worm_best, worm_plant_best,
                              by=c("subject_id"="query_id"),
                              suffix = c(".plant_worm", ".worm_plant"))
head(plant_with_recip)
```




** Exercise X**

Reciprocal best hits...

Venn Diagram...

Relationship between E-value and score

**Excercise 10**:
It sometimes is useful to create a random subset of a large data set like the fly.worm results.  Create a new dataframe `fly.worm.small` that has 1,000 rows taken at random from the larger data set.  Hint: one way is to use the `sample()` function to create a vector of rownumbers to keep and then use the `[ ]` extractor.




## Intro to plotting
One of the strengths of R is its graphics.  We are going to take a break from the bioinformatics datasets and work on a tutorial made by the 2014 TA Ciera.  This dataset uses the [stereotypes.csv file](http://cierareports.org/downloads/r_tutorial/stereotypes.csv) which you can download by clicking on link.

Ciera describes the dataset as follows: "799 individuals were surveyed from ages 18 to 26 about their weekly their consumption habits. They were further categorized based on their clothing and music choices and put into four stereotypes: nerds, hipsters, metalheads, and hippies. Consumption of coffee (per cup), computer usage (hourly), showers, beers(per 12oz), and tacos. Oh yeah, this dataset is fake."

Before beginning, import the "stereotypes.csv" into an R object called "stereotypes"

**Exercise 11**:
Use the `summary()` and `head()` functions to examine the stereotypes object.

### the subset command
In `swirl` you learned how to subset objects using square brackets [].  An alternative method is to use the `subset()` command.  To learn about the subset command, go to [Ciera's tutorial](http://cierareports.org/blog/2013/10/18/rCourse2013/index.html) and scroll down until you get to the part on subset() (It is near "Question 1".  Work through the tutorial section on subsetting.  Record the answers to your "activity breaks" in your R Notebook.

### Plotting tutorial
Now that you know about `subsetting`, it is time to follow the plotting tutorial.

Please follow the __visualization__ part of [Ciera's tutorial](http://cierareports.org/blog/2013/10/18/rCourse2013/index.html#section2).  You can stop when you get to "Basic Statistics".  Record the answers to the "activity break sections" in your R markdown file.

Remember to save and commit your Rmarkdown file.

### Visualization of the BLAST data set
Now lets apply what we have learned to the BLAST data set.

**Exercise 12**
Use ggplot to explore the relationship between score ("Score") and alignment length("len").  Provide a plot that illustrates the relationship and describe the relationship in words.

**Excercise 13**
While you might expect that BLAST results with long alignments would have high scores, this is not always the case.  Form a hypothesis as to what might influence the relationship between alignment length and score.  Use ggplot to make a new plot to explore this hypothesis.  Does the plot support your hypothesis or not?  State your hypothesis, provide the code for your plot, and state your conclusion.

** Turning in the assignment**

* Click the "Preview" Button to generate an up-to-date html version of your notebook.  Check it to make sure you are happy with its content.
* add your .Rmd and .nb.html files to the repository and commit the changes.
* push the changes
