---
title: "Rice GWAS"
layout: lab
tags:
- R
- SNPs
hidden: true
---

Today we will analyze variation in the phenotypic data.  We specifically want to know if various phenotypes vary by region or population, and whether we can identify SNPs that are associated with variation in a trait of interest.

### Goals

1. Examine and summarize the root Aluminum tolerance data
2. Perform a GWAS for Aluminum tolerance
3. Find candidate genes underlying a GWAS peak.

## Plotting the Amylose data

We need to get the population assignments from PSmix in the same data frame as the MDS genotype assignments and the phenotypic data.  You should have saved this at the end of Tuesday's lab



```r
load("data_from_SNP_lab.RData")
```

check it

```r
head(geno.mds)
```

```
##                V1         V2 popID           pr1           pr2
## NSFTV1  17.118285   1.961496     4 4.940656e-324 4.940656e-324
## NSFTV3 -23.709825  10.237016     1  1.000000e+00 4.227075e-115
## NSFTV4 -18.333071 -11.214147     2  1.040618e-01  8.944053e-01
## NSFTV5   2.449566  -4.488631     3  2.322312e-06  3.209291e-01
## NSFTV6 -17.708530 -11.116512     2  5.048234e-02  9.011619e-01
## NSFTV7   9.250797  -1.317508     3  1.042226e-18  4.515324e-02
##                  pr3           pr4
## NSFTV1  1.435086e-76  1.000000e+00
## NSFTV3 5.671628e-234 5.543975e-274
## NSFTV4  1.532923e-03  1.035332e-54
## NSFTV5  4.212880e-01  2.577806e-01
## NSFTV6  4.835579e-02  2.491382e-32
## NSFTV7  9.108892e-01  4.395752e-02
```

```r
## If your geno.mds file does not have the columns indicated above, then go back to Tuesday's lab, and re-make it.
```

Do the merge

```r
data.pheno.mds <- merge(geno.mds,data.pheno,by="row.names",all=T) #even if you already have this object, rerun it so that the AmID and pr1-4 columns are brought in
```


Lets examine the distribution of Amylose values.  We can tell qplot to produce a histogram as follows:

```r
library(ggplot2)
qplot(x=Amylose.content,data=data.pheno.mds,geom="histogram")
```

It might be interesting to ask if the distributions look similar for different regions.
We can also easily produce separate histograms for each level of a factor.
Because this is getting to be a more complex plot we will use ggplot to build up the plot.  First we create a plot object, called pl, using the ggplot() function.  Then we add additional information to the plot.  the mapping=aes() argument tells R about the "aesthetics" of the plot, in other words which variable should be mapped to which aspect of the plot.

```r
pl <- ggplot(data=data.pheno.mds,aes(x=Amylose.content)) #create the basic plot object
pl <- pl + geom_histogram(binwidth=3) #tell R that we want a histogram, with binwidth of 3
pl <- pl + facet_wrap(facets= ~ Region, ncol=3) # a separate plot ("facet") for each region, arranged in 3 columns
pl <- pl + ggtitle("Amylose Content") #add a title
pl #display the plot
```
We are a little sparse on data for histograms to be a great way to visualize the results.

An alternative and useful way to display the data is to use a [boxplot](http://en.wikipedia.org/wiki/Box_plot). The horizontal line is the median and the box is the inter-quartile range (25th to 75th percentile).  See the link for more details.

```r
qplot(x=Region,y=Amylose.content,geom="boxplot",data=data.pheno.mds)
```

For the rest of the lab, choose a trait to work on.  Alu.Tol, Protein.content, Seed.volume, and Seed.number.per.panicle all will work.  (Others might also but I haven't looked at them all).

If you choose Alu.Tol, The "Alu.Tol" column indicates the ratio of root growth  with aluminum to root growth without Aluminum.  A value of 1 means equal growth with or without Aluminum.  A value of 0.5 means that the roots grew half as well on Aluminum than without Aluminum.


**Exercise 5:** 

* Plot your chosen trait data
  * as a **single histogram** for all of the data
  * as **separate histograms** for each of the 4 population assignments made by PSMix
  * as a **boxplot** separated by population. 
* Based on these histograms do you think that your trait varies by population? 
* **BONUS** Try using the "violin" geom.  What is this showing?
  
*Hint:* you will need to use a different binwidth than I used with Amylose (or don't specify it at all and let R choose the default).

*Hint:* the relevant column names for population are "popID".  


These plots may make you want to calculate the mean for your trait. We can use the `mean()` function to obtain the mean of a column:

```r
mean(data.pheno.mds$Amylose.content,na.rm=T) #the na.rm argument tells R to ignore missing data coded by "NA"
```

More interestingly we can use `tapply()` to apply a function to each subset of the data.  

```r
tapply(X=data.pheno.mds$Amylose.content,INDEX=data.pheno.mds$Region,FUN=min,na.rm=T)
```
The first argument to `tapply()` is the data we want to summarize, the second argument is how the data should be split, the third argument is the function we want to apply (`min` in this case) and the fourth argument are additional arguments to be passed to the function listed in the argument.

For Amylose, there appear to be differences in the means between population groups.  We can ask if this is significant by performing an ANOVA.

```r
aov1 <- aov(Amylose.content ~ Region,data=data.pheno.mds) #1-way ANOVA for Amylose.content by Region
summary(aov1)
```
The very low p-value for Region indicates that amylose content varies significantly by region.

**Exercise 6:** 
* Obtain the mean of your trait for each of the 4 PSMix populations.
* Perform an ANOVA for your trait to test if it varies significantly by population.  Show your code, the ANOVA output, and provide an interpretation.
* Discuss: Do your results suggest a problem for GWAS?



```r
aov2 <- aov(Alu.Tol ~ popID,data=data.pheno.mds)
summary(aov2)
```

## GWAS


Finally we are ready to perform a GWAS.  Please run the commands below to set up the data for analysis.

```r
#load information about snp chromosome and position
snpinfo <- read.csv("snpInfo.csv",row.names=1)

head(snpinfo) #note one column for chromosome and one for position (in base pairs)

# you will need the data.geno file from the previous lab.  If you don't have it in your workspace, reload it
data.geno <- read.csv("Rice_44K_genotypes.csv", row.names=1, na.strings=c("NA","00")) 

#next merge the genotype information with the phenotype and population info:
head(data.pheno.mds) #note: need to get rownames assigned correctly again (they were lost after the merge)
rownames(data.pheno.mds) <- data.pheno.mds$Row.names
data.geno.pheno <- merge(data.pheno.mds,data.geno,by="row.names")
#you can ignore the warning 
```

The associations will be done with the package [SNPassoc](http://bioinformatics.oxfordjournals.org/content/23/5/654.2.long).  The  commands below create objects for SNPassoc to use.  

In the interest of time we will only examine chromosome 3.


```r
library(SNPassoc) #load the package that does the associations
#if you get an error that the package is not available, use install.packages("SNPassoc") to install it.

#create new data frames containing only chromosome 3 information.
#grep() is the R version of the linux grep command that you saw in Ian's section.  So the command below extracts columns 1:47 and all chromosome 3 genotype columns
data.geno.pheno3 <- data.geno.pheno[,c(1:47,grep("X3_",colnames(data.geno.pheno)))]
snpinfo3 <- snpinfo[snpinfo$chr==3,]

#convert SNPinfo to a format that SNPassoc can use
snps3 <- setupSNP(data.geno.pheno3,48:ncol(data.geno.pheno3),sort=T,info=snpinfo3,sep="")
```
For the actual GWAS we will compare a model with and without correction for population structure.  The correction method that we use, including ancestral genome proportions as co-variates, is a good first step.  However it is often better to include a Kinship matrix of genetic relatedness.  Since this tutorial was created an R package with state-of-the-art methods has been released.  The package is GAPIT [website](http://www.maizegenetics.net/#!gapit/cmkv) and [paper](http://bioinformatics.oxfordjournals.org/content/28/18/2397.long).  If you are going to do a serious GWAS you should consider using GAPIT.

We use a co-dominant model of allelic affects (allowing each allele to act additively).  You can experiment with other models if you would like.

Substitute your chosen trait for "Alu.Tol" below if you are not working on Aluminum Tolerance.


```r
#analysis without population structure correction
#this takes ~ 5 minutes to run.
wg3 <- WGassociation(Alu.Tol,data=snps3,model="co",genotypingRate=50)
plot(wg3,print.label.SNPs=FALSE)

#the p-values for the co-dominant model are extracted by using the codominant() function 
#determine the number of significant SNPs (p < 0.00001):
sum(codominant(wg3) < 1e-5)
```

Population stratification can lead to an inflated number of false positives.  We can ask if this is occurring in our data by comparing the observed distribution of p-values against the expected distribution of p-values assuming no GWAS hits.  Because we expect the VAST majority of the SNPS to not have an association, the expected and observed distributions should match except for maybe a few SNPs at the very tail.  This is a modified version of a quantile-quantile (qq) plot.  We can make such a plot as shown below

```r
qqpval(codominant(wg3)) #remember that codominant(wg3) returns the observed p-values.
```

The blue points compare the observed and expected p-values.  The diagonal line shows where these points would be if the two distributions were equal.  Do you think there is evidence of population structure confounding our result?

We can begin to correct for population structure by including the predicted population membership from Tuesday's lab as co-factors in the analysis.

```r
#analysis with population structure correction:
wg3.corrected <- WGassociation(Alu.Tol ~ pr1 + pr2 + pr3 + pr4,data=snps3,model="co",genotypingRate=50)
#analysis with population structure correction.
plot(wg3.corrected,print.label.SNPs=FALSE)
sum(codominant(wg3.corrected) < 1e-5)
```

Now take a look at the qqplot for the corrected analysis:

```r
qqpval(codominant(wg3.corrected))
```
Better, but still the p-values are inflated.  The next step would be to create and include a kinship matrix and run some more sophisticated analyses in [GAPIT](http://www.maizegenetics.net/#!gapit/cmkv).  Unfortunately that is outside the scope of this lab.  Still the most significant SNPs may very well be real.

**Exercise 7:** Describe and discuss the differences in the analysis with and without population structure correction.  Which do you think is the better one to follow-up on, and why?

Next we want to know what SNPs were significant @ p < 0.00001.  We can do that as follows:

```r
#use the square bracked extractions command to extract all rows where the SNP p-value is less than 1e-5
snpinfo3[codominant(wg3.corrected) < 1e-5,]

#if we want to add the pvals to the output:
cbind(snpinfo3[codominant(wg3.corrected) < 1e-5,],codominant(wg3.corrected)[codominant(wg3.corrected) < 1e-5])
```

**Exercise 8:** Look for genes close to your SNP at the [rice genome browser](http://rice.plantbiology.msu.edu/cgi-bin/gbrowse/rice/).  Pick a significant SNP from your analysis and enter its chromosome and position in the search box.  The browser wants you to enter a start and stop position, so for example, you should enter "Chr3:30449857..30449857" and then choose "show 20kb" from the pulldown menu on the left hand side.  Report the SNP you chose and the three closest genes.  These are candidate genes for determining the phenotype of your trait of interest in the rice population.  Briefly discuss these genes as possible candidates for the GWAS peak.  __Include a Screenshot of the genome browser in your answer__




