---
title: "Illumina continued: analyze VCF file in R"
layout: lab
tags:
- R
- Brassica
- Illumina
hidden: true
output: html_document
---

In Thursday's lab we finished by making a .vcf file.  I realized that we need to add a couple of additional flags to `samtools mpileup` to get all of the information that we need about these polymorphisms (Thursday's lab is now updated).  Unless you have already run `samtools mpileup` with the -DV flags, please re-run it as below.  Or [download a new version]({{site.baseurl}}/data/IMB211_R500.vcf).

```
samtools mpileup -DVuf ../Brapa_reference/BrapaV1.5_chrom_only.fa IMB211_rmdup.bam R500_rmdup.bam | bcftools view -vcg - > IMB211_R500.vcf
``` 

`Samtools mpileup` calculates the number of reference and alternate alleles at each position in the genome and genotype likelihoods.  `bcftools` makes a call of the most likely genotype.

You should check the man page for the meaning of the flags, but briefly, 

* for samtools, 
    * -D and -V indicate that per sample depth and variant depth should be reported.  
    * -u is uncompressed format
    * -f specifies the reference fasta file.  
* For bcftools,
    * -c specifies that variants should be called. 
    * -v limits the output to variant sites only
    * -g reports separate genotype calls for each sample
    
The .vcf file contains information on polymorphisms.  It is a tab-delimited file that is easily imported into R.  We want to filter the SNPs to remove low quality SNPs.  We also could use this for downstream analyses, i.e. use these as genetic markers for mapping or QTL mapping, look for genes with coding changes, etc.


```r
library(ggplot2)
vcf.data <- read.table("IMB211_R500.vcf",as.is=T)
head(vcf.data)
```


```
##    V1   V2 V3   V4 V5     V6 V7
## 1 A01 4740  . TACA TA  12.70  .
## 2 A01 4862  .    G  T 999.00  .
## 3 A01 4932  .    T TA   4.42  .
## 4 A01 5070  .    G  A 999.00  .
## 5 A01 5245  .    A  G 152.00  .
## 6 A01 5313  .    G  C  10.40  .
##                                                                                                      V8
## 1                                       INDEL;IS=1,0.111111;DP=2;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 2                                        DP=55;VDB=1.108038e-01;AF1=1;AC1=4;DP4=0,0,28,27;MQ=50;FQ=-107
## 3                                       INDEL;IS=1,0.037037;DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 4                                        DP=62;VDB=4.078190e-01;AF1=1;AC1=4;DP4=0,0,30,32;MQ=43;FQ=-101
## 5 DP=30;VDB=1.424752e-01;RPB=-4.445004e-01;AF1=0.2513;AC1=1;DP4=8,10,4,8;MQ=46;FQ=153;PV4=0.71,1,0.17,1
## 6                                                           DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-27.4
##               V9                   V10                    V11
## 1 GT:PL:DP:DV:GQ      1/1:50,3,0:1:1:4        0/1:0,0,0:0:0:3
## 2 GT:PL:DP:DV:GQ 1/1:255,75,0:25:25:99  1/1:255,90,0:30:30:99
## 3 GT:PL:DP:DV:GQ       0/1:0,0,0:0:0:3       0/1:40,3,0:1:1:4
## 4 GT:PL:DP:DV:GQ 1/1:255,69,0:23:23:99 1/1:255,117,0:39:39:99
## 5 GT:PL:DP:DV:GQ   0/0:0,21,216:7:0:23 0/1:185,0,192:23:12:99
## 6 GT:PL:DP:DV:GQ      1/1:40,3,0:1:1:4        0/1:0,0,0:0:0:3
```

There is no column heading in what we imported.  There is a column heading in the .vcf file but is not loaded into R because its first character is "#".  Here we get it by using a `system()` call.  `system()` causes R to issue a command to the Linux shell.  The column headings in the .vcf start with "#CHROM" so we use Linux's version of `grep` to search for that line.  The argument `intern = TRUE` tells R that we want to capture the output of the command.


```r
vcf.header <- system("grep '#C' IMB211_R500.vcf",intern = TRUE) #might not work on Windows
vcf.header
vcf.header <- sub("#","",vcf.header) #get rid of the pound sign
```


```
## [1] "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tIMB211_rmdup.bam\tR500_rmdup.bam"
```

The header is tab delimited, so we split on tabs and use those as column names

```r
vcf.header <- unlist(strsplit(vcf.header,split="\t"))
colnames(vcf.data) <- vcf.header
head(vcf.data)
```

```
##   CHROM  POS ID  REF ALT   QUAL FILTER
## 1   A01 4740  . TACA  TA  12.70      .
## 2   A01 4862  .    G   T 999.00      .
## 3   A01 4932  .    T  TA   4.42      .
## 4   A01 5070  .    G   A 999.00      .
## 5   A01 5245  .    A   G 152.00      .
## 6   A01 5313  .    G   C  10.40      .
##                                                                                                    INFO
## 1                                       INDEL;IS=1,0.111111;DP=2;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 2                                        DP=55;VDB=1.108038e-01;AF1=1;AC1=4;DP4=0,0,28,27;MQ=50;FQ=-107
## 3                                       INDEL;IS=1,0.037037;DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 4                                        DP=62;VDB=4.078190e-01;AF1=1;AC1=4;DP4=0,0,30,32;MQ=43;FQ=-101
## 5 DP=30;VDB=1.424752e-01;RPB=-4.445004e-01;AF1=0.2513;AC1=1;DP4=8,10,4,8;MQ=46;FQ=153;PV4=0.71,1,0.17,1
## 6                                                           DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-27.4
##           FORMAT      IMB211_rmdup.bam         R500_rmdup.bam
## 1 GT:PL:DP:DV:GQ      1/1:50,3,0:1:1:4        0/1:0,0,0:0:0:3
## 2 GT:PL:DP:DV:GQ 1/1:255,75,0:25:25:99  1/1:255,90,0:30:30:99
## 3 GT:PL:DP:DV:GQ       0/1:0,0,0:0:0:3       0/1:40,3,0:1:1:4
## 4 GT:PL:DP:DV:GQ 1/1:255,69,0:23:23:99 1/1:255,117,0:39:39:99
## 5 GT:PL:DP:DV:GQ   0/0:0,21,216:7:0:23 0/1:185,0,192:23:12:99
## 6 GT:PL:DP:DV:GQ      1/1:40,3,0:1:1:4        0/1:0,0,0:0:0:3
```

Now with the headers attached the file makes a bit more sense.  Detailed information [is here](https://samtools.github.io/hts-specs/VCFv4.2.pdf), but briefly:

* CHROM and POS should be obvious
* ID is not used here
* REF is the reference sequence (From the fasta file)
* ALT is the allele found in out sample(s)
* QUAL is the phred-scaled quality of the polymorphism.
* FILTER is not used here.  There are programs that will place flags in this field to indicate that a SNP should be filtered out.
* INFO lots of INFO.  We can learn about these from the beginning of the vcf file (See below)
* FORMAT tells is the identity of the numbers in the next two fields
    * GT: The most probably genotype.  1/1 is homozygous alternate; 0/0 is homozygous reference; 0/1 is heterozygous.
    * PL: Phred scaled relative likelihoods of the 3 genotypic possibilities (0/0 , 0/1, 1/1).  The most likely is scored "0".  Higher is worse.
    * DP: Total high quality read counts
    * DV: High quality read counts that match the alternate allele
    * GQ: Overall Phred-scaled Genotype Quality; higher is better (probability that the genotype call is wrong)

This information is at the beginning of the .vcf file which we could look at with `less` at the Linux command line, or we can look at it in R using system calls.  (Might not work on Windows)


```r
system("grep '##INFO' IMB211_R500.vcf")

system("grep '##FORMAT' IMB211_R500.vcf")
```


```
##  [1] "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">"                                                                                                  
##  [2] "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases\">"                               
##  [3] "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Root-mean-square mapping quality of covering reads\">"                                                              
##  [4] "##INFO=<ID=FQ,Number=1,Type=Float,Description=\"Phred probability of all samples being the same\">"                                                                   
##  [5] "##INFO=<ID=AF1,Number=1,Type=Float,Description=\"Max-likelihood estimate of the first ALT allele frequency (assuming HWE)\">"                                         
##  [6] "##INFO=<ID=AC1,Number=1,Type=Float,Description=\"Max-likelihood estimate of the first ALT allele count (no HWE assumption)\">"                                        
##  [7] "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">"                                                                     
##  [8] "##INFO=<ID=IS,Number=2,Type=Float,Description=\"Maximum number of reads supporting an indel and fraction of indel reads\">"                                           
##  [9] "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">"                                      
## [10] "##INFO=<ID=G3,Number=3,Type=Float,Description=\"ML estimate of genotype frequencies\">"                                                                               
## [11] "##INFO=<ID=HWE,Number=1,Type=Float,Description=\"Chi^2 based HWE test P-value based on G3\">"                                                                         
## [12] "##INFO=<ID=CLR,Number=1,Type=Integer,Description=\"Log ratio of genotype likelihoods with and without the constraint\">"                                              
## [13] "##INFO=<ID=UGT,Number=1,Type=String,Description=\"The most probable unconstrained genotype configuration in the trio\">"                                              
## [14] "##INFO=<ID=CGT,Number=1,Type=String,Description=\"The most probable constrained genotype configuration in the trio\">"                                                
## [15] "##INFO=<ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">"                                           
## [16] "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">"                                                                         
## [17] "##INFO=<ID=PC2,Number=2,Type=Integer,Description=\"Phred probability of the nonRef allele frequency in group1 samples being larger (,smaller) than in group2.\">"     
## [18] "##INFO=<ID=PCHI2,Number=1,Type=Float,Description=\"Posterior weighted chi^2 P-value for testing the association between group1 and group2 samples.\">"                
## [19] "##INFO=<ID=QCHI2,Number=1,Type=Integer,Description=\"Phred scaled PCHI2.\">"                                                                                          
## [20] "##INFO=<ID=PR,Number=1,Type=Integer,Description=\"# permutations yielding a smaller PCHI2.\">"                                                                        
## [21] "##INFO=<ID=QBD,Number=1,Type=Float,Description=\"Quality by Depth: QUAL/#reads\">"                                                                                    
## [22] "##INFO=<ID=RPB,Number=1,Type=Float,Description=\"Read Position Bias\">"                                                                                               
## [23] "##INFO=<ID=MDV,Number=1,Type=Integer,Description=\"Maximum number of high-quality nonRef reads in samples\">"                                                         
## [24] "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias (v2) for filtering splice-site artefacts in RNA-seq data. Note: this version may be broken.\">"
```

```
## [1] "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">"                                       
## [2] "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">"                              
## [3] "##FORMAT=<ID=GL,Number=3,Type=Float,Description=\"Likelihoods for RR,RA,AA genotypes (R=ref,A=alt)\">"
## [4] "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"# high-quality bases\">"                          
## [5] "##FORMAT=<ID=DV,Number=1,Type=Integer,Description=\"# high-quality non-reference bases\">"            
## [6] "##FORMAT=<ID=SP,Number=1,Type=Integer,Description=\"Phred-scaled strand bias P-value\">"              
## [7] "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">"
```

To be able to use the SNP data in R we need to break up the data in the final two columns.  We use the command `strsplit()` for this.  `strsplit()` splits a character string based at every occurrence of a character.  We will make a temporary matrix to hold the split data and then bind it back to the original data file.


```r
IMB211.tmp <- matrix(
  unlist(strsplit(vcf.data$IMB211_rmdup.bam,split = ":")),
  nrow=nrow(vcf.data),
  byrow=TRUE
  )
head(IMB211.tmp)
```

```
##      [,1]  [,2]       [,3] [,4] [,5]
## [1,] "1/1" "50,3,0"   "1"  "1"  "4" 
## [2,] "1/1" "255,75,0" "25" "25" "99"
## [3,] "0/1" "0,0,0"    "0"  "0"  "3" 
## [4,] "1/1" "255,69,0" "23" "23" "99"
## [5,] "0/0" "0,21,216" "7"  "0"  "23"
## [6,] "1/1" "40,3,0"   "1"  "1"  "4"
```

```r
colnames(IMB211.tmp) <- paste("IMB211",c("gt","lik","tot.depth","alt.depth","qual"),sep="_")

R500.tmp <- matrix(
  unlist(strsplit(vcf.data$R500_rmdup.bam,split = ":")),
  nrow=nrow(vcf.data),
  byrow=TRUE
  )
head(R500.tmp)
```

```
##      [,1]  [,2]        [,3] [,4] [,5]
## [1,] "0/1" "0,0,0"     "0"  "0"  "3" 
## [2,] "1/1" "255,90,0"  "30" "30" "99"
## [3,] "0/1" "40,3,0"    "1"  "1"  "4" 
## [4,] "1/1" "255,117,0" "39" "39" "99"
## [5,] "0/1" "185,0,192" "23" "12" "99"
## [6,] "0/1" "0,0,0"     "0"  "0"  "3"
```

```r
colnames(R500.tmp) <- paste("R500",c("gt","lik","tot.depth","alt.depth","qual"),sep="_")

vcf.data <- cbind(vcf.data,IMB211.tmp,R500.tmp,stringsAsFactors=FALSE)
summary(vcf.data)
```

```
##     CHROM                POS                ID           
##  Length:73485       Min.   :    4740   Length:73485      
##  Class :character   1st Qu.: 5028712   Class :character  
##  Mode  :character   Median :10714393   Mode  :character  
##                     Mean   :12673284                     
##                     3rd Qu.:21307519                     
##                     Max.   :26785665                     
##      REF                ALT                 QUAL           FILTER         
##  Length:73485       Length:73485       Min.   :  3.01   Length:73485      
##  Class :character   Class :character   1st Qu.: 32.00   Class :character  
##  Mode  :character   Mode  :character   Median :136.00   Mode  :character  
##                                        Mean   :336.60                     
##                                        3rd Qu.:999.00                     
##                                        Max.   :999.00                     
##      INFO              FORMAT          IMB211_rmdup.bam  
##  Length:73485       Length:73485       Length:73485      
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##  R500_rmdup.bam      IMB211_gt          IMB211_lik       
##  Length:73485       Length:73485       Length:73485      
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##  IMB211_tot.depth   IMB211_alt.depth   IMB211_qual       
##  Length:73485       Length:73485       Length:73485      
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##    R500_gt            R500_lik         R500_tot.depth    
##  Length:73485       Length:73485       Length:73485      
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##  R500_alt.depth      R500_qual        
##  Length:73485       Length:73485      
##  Class :character   Class :character  
##  Mode  :character   Mode  :character  
##                                       
##                                       
## 
```

There is one problem: we need to convert the columns that should be numbers back to numeric.  This is done by extracting the relevant columns, using `apply` to apply the `as.numeric` function to each column, and then placing the results back in the correct place


```r
head(vcf.data)
```

```
##   CHROM  POS ID  REF ALT   QUAL FILTER
## 1   A01 4740  . TACA  TA  12.70      .
## 2   A01 4862  .    G   T 999.00      .
## 3   A01 4932  .    T  TA   4.42      .
## 4   A01 5070  .    G   A 999.00      .
## 5   A01 5245  .    A   G 152.00      .
## 6   A01 5313  .    G   C  10.40      .
##                                                                                                    INFO
## 1                                       INDEL;IS=1,0.111111;DP=2;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 2                                        DP=55;VDB=1.108038e-01;AF1=1;AC1=4;DP4=0,0,28,27;MQ=50;FQ=-107
## 3                                       INDEL;IS=1,0.037037;DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 4                                        DP=62;VDB=4.078190e-01;AF1=1;AC1=4;DP4=0,0,30,32;MQ=43;FQ=-101
## 5 DP=30;VDB=1.424752e-01;RPB=-4.445004e-01;AF1=0.2513;AC1=1;DP4=8,10,4,8;MQ=46;FQ=153;PV4=0.71,1,0.17,1
## 6                                                           DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-27.4
##           FORMAT      IMB211_rmdup.bam         R500_rmdup.bam IMB211_gt
## 1 GT:PL:DP:DV:GQ      1/1:50,3,0:1:1:4        0/1:0,0,0:0:0:3       1/1
## 2 GT:PL:DP:DV:GQ 1/1:255,75,0:25:25:99  1/1:255,90,0:30:30:99       1/1
## 3 GT:PL:DP:DV:GQ       0/1:0,0,0:0:0:3       0/1:40,3,0:1:1:4       0/1
## 4 GT:PL:DP:DV:GQ 1/1:255,69,0:23:23:99 1/1:255,117,0:39:39:99       1/1
## 5 GT:PL:DP:DV:GQ   0/0:0,21,216:7:0:23 0/1:185,0,192:23:12:99       0/0
## 6 GT:PL:DP:DV:GQ      1/1:40,3,0:1:1:4        0/1:0,0,0:0:0:3       1/1
##   IMB211_lik IMB211_tot.depth IMB211_alt.depth IMB211_qual R500_gt
## 1     50,3,0                1                1           4     0/1
## 2   255,75,0               25               25          99     1/1
## 3      0,0,0                0                0           3     0/1
## 4   255,69,0               23               23          99     1/1
## 5   0,21,216                7                0          23     0/1
## 6     40,3,0                1                1           4     0/1
##    R500_lik R500_tot.depth R500_alt.depth R500_qual
## 1     0,0,0              0              0         3
## 2  255,90,0             30             30        99
## 3    40,3,0              1              1         4
## 4 255,117,0             39             39        99
## 5 185,0,192             23             12        99
## 6     0,0,0              0              0         3
```

```r
vcf.data[,c("POS","QUAL","IMB211_tot.depth","IMB211_alt.depth","IMB211_qual","R500_tot.depth","R500_alt.depth","R500_qual")] <- 
  apply(vcf.data[,c("POS","QUAL","IMB211_tot.depth","IMB211_alt.depth","IMB211_qual","R500_tot.depth","R500_alt.depth","R500_qual")],
        2,
        as.numeric
        )
head(vcf.data)
```

```
##   CHROM  POS ID  REF ALT   QUAL FILTER
## 1   A01 4740  . TACA  TA  12.70      .
## 2   A01 4862  .    G   T 999.00      .
## 3   A01 4932  .    T  TA   4.42      .
## 4   A01 5070  .    G   A 999.00      .
## 5   A01 5245  .    A   G 152.00      .
## 6   A01 5313  .    G   C  10.40      .
##                                                                                                    INFO
## 1                                       INDEL;IS=1,0.111111;DP=2;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 2                                        DP=55;VDB=1.108038e-01;AF1=1;AC1=4;DP4=0,0,28,27;MQ=50;FQ=-107
## 3                                       INDEL;IS=1,0.037037;DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-34.9
## 4                                        DP=62;VDB=4.078190e-01;AF1=1;AC1=4;DP4=0,0,30,32;MQ=43;FQ=-101
## 5 DP=30;VDB=1.424752e-01;RPB=-4.445004e-01;AF1=0.2513;AC1=1;DP4=8,10,4,8;MQ=46;FQ=153;PV4=0.71,1,0.17,1
## 6                                                           DP=1;AF1=1;AC1=4;DP4=0,0,1,0;MQ=50;FQ=-27.4
##           FORMAT      IMB211_rmdup.bam         R500_rmdup.bam IMB211_gt
## 1 GT:PL:DP:DV:GQ      1/1:50,3,0:1:1:4        0/1:0,0,0:0:0:3       1/1
## 2 GT:PL:DP:DV:GQ 1/1:255,75,0:25:25:99  1/1:255,90,0:30:30:99       1/1
## 3 GT:PL:DP:DV:GQ       0/1:0,0,0:0:0:3       0/1:40,3,0:1:1:4       0/1
## 4 GT:PL:DP:DV:GQ 1/1:255,69,0:23:23:99 1/1:255,117,0:39:39:99       1/1
## 5 GT:PL:DP:DV:GQ   0/0:0,21,216:7:0:23 0/1:185,0,192:23:12:99       0/0
## 6 GT:PL:DP:DV:GQ      1/1:40,3,0:1:1:4        0/1:0,0,0:0:0:3       1/1
##   IMB211_lik IMB211_tot.depth IMB211_alt.depth IMB211_qual R500_gt
## 1     50,3,0                1                1           4     0/1
## 2   255,75,0               25               25          99     1/1
## 3      0,0,0                0                0           3     0/1
## 4   255,69,0               23               23          99     1/1
## 5   0,21,216                7                0          23     0/1
## 6     40,3,0                1                1           4     0/1
##    R500_lik R500_tot.depth R500_alt.depth R500_qual
## 1     0,0,0              0              0         3
## 2  255,90,0             30             30        99
## 3    40,3,0              1              1         4
## 4 255,117,0             39             39        99
## 5 185,0,192             23             12        99
## 6     0,0,0              0              0         3
```

```r
summary(vcf.data)
```

```
##     CHROM                POS                ID           
##  Length:73485       Min.   :    4740   Length:73485      
##  Class :character   1st Qu.: 5028712   Class :character  
##  Mode  :character   Median :10714393   Mode  :character  
##                     Mean   :12673284                     
##                     3rd Qu.:21307519                     
##                     Max.   :26785665                     
##      REF                ALT                 QUAL           FILTER         
##  Length:73485       Length:73485       Min.   :  3.01   Length:73485      
##  Class :character   Class :character   1st Qu.: 32.00   Class :character  
##  Mode  :character   Mode  :character   Median :136.00   Mode  :character  
##                                        Mean   :336.60                     
##                                        3rd Qu.:999.00                     
##                                        Max.   :999.00                     
##      INFO              FORMAT          IMB211_rmdup.bam  
##  Length:73485       Length:73485       Length:73485      
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##  R500_rmdup.bam      IMB211_gt          IMB211_lik        IMB211_tot.depth
##  Length:73485       Length:73485       Length:73485       Min.   :  0.00  
##  Class :character   Class :character   Class :character   1st Qu.:  1.00  
##  Mode  :character   Mode  :character   Mode  :character   Median :  5.00  
##                                                           Mean   : 13.04  
##                                                           3rd Qu.: 17.00  
##                                                           Max.   :186.00  
##  IMB211_alt.depth   IMB211_qual      R500_gt            R500_lik        
##  Min.   :  0.000   Min.   : 2.00   Length:73485       Length:73485      
##  1st Qu.:  0.000   1st Qu.: 5.00   Class :character   Class :character  
##  Median :  2.000   Median :20.00   Mode  :character   Mode  :character  
##  Mean   :  8.303   Mean   :36.15                                        
##  3rd Qu.:  9.000   3rd Qu.:64.00                                        
##  Max.   :179.000   Max.   :99.00                                        
##  R500_tot.depth   R500_alt.depth      R500_qual    
##  Min.   :  0.00   Min.   :  0.000   Min.   : 2.00  
##  1st Qu.:  1.00   1st Qu.:  0.000   1st Qu.: 6.00  
##  Median :  5.00   Median :  2.000   Median :21.00  
##  Mean   : 13.56   Mean   :  9.367   Mean   :37.12  
##  3rd Qu.: 17.00   3rd Qu.: 11.000   3rd Qu.:66.00  
##  Max.   :188.00   Max.   :183.000   Max.   :99.00
```

Now let's do some data exploration.  First let's limit ourselves to the SNP data.  In the line below, `grepl` returns a TRUE or FALSE depending on whether or not "INDEL" is found in the info field.  By reversing this (using "!") we can only keep those rows without an INDEL, (i.e. we keep the SNPs) 

```r
snp.data <- vcf.data[!grepl("INDEL",vcf.data$INFO),]
```

__Practice 1__  
To explore the quality of our data, make two histograms, one for R500 genotype quality and one for IMB211 quality


__Practice 2__  
__a__Subset the data to only keep positions where both IMB211 and R500 have a quality >= 20.  
__b__ At a quality of 20 what is the probability that the genotype call is wrong? (Hint: check the Wikipedia entry on Phred scores)


We can count the number of homozygous and heterozygous snps using table:


```r
table(snp.data.good$IMB211_gt)
```

```
## 
##   0/0   0/1   1/1 
##  6695  1346 20631
```

```r
table(snp.data.good$R500_gt)
```

```
## 
##   0/0   0/1   1/1 
##  5420  1292 21960
```

We can even count the numbers common and unique to each genotype


```r
ftable(snp.data.good[,c("IMB211_gt","R500_gt")])
```

```
##           R500_gt   0/0   0/1   1/1
## IMB211_gt                          
## 0/0                   0   497  6198
## 0/1                 519   701   126
## 1/1                4901    94 15636
```

__Practice 3__  
__a__ Using the higher quality SNP list, plot the position along the chromosome of each SNP (x axis), read depth (R500 and IMB211 combined) (y axis).  
__b__ Why might there be gaps with no SNPs?  
__Bonus__: color each SNP for whether it is R500 specific, IMB specific, or common.


__For Fun (?)--not graded--__  
Plot the number of each type of base change (A->G, etc).  Are there differences?  Is this expected?
