---
layout: lab
title: BLAST for paralogs and orthologs
hidden: true    <!-- To prevent post from being displayed as regular blog post -->
tags:
- Linux
- sequence
---



## Part 1: Orthologs ##

In the last lab, you learned that

* There is always a best local alignment, even in random sequences
* The distribution of random alignment scores is not normal
* The random background depends on
	* The lengths of the sequences
	* The composition of the sequences
	* The scoring parameters (matrix, gaps)
* The significance of a score is difficult to assess

We previously called the best match for the C.elegans protein B0213.10 in the A. thaliana and D.
melanogaster proteomes the ortholog. but the ortholog is operationally
defined as the best reciprocal match between proteomes. That is, after
finding the best match to the plant proteome, one must take the plant protein
and search it against the worm proteome to determine if it finds
B0213.10. If if it does, the proteins are considered orthologous.

Let's do something ambitious. Let's find ALL the orthologous proteins
between worm and plant. How long will it take to search every protein
against every other protein? You made this calculation in Lab 1 with the
`water` program and it was outrageously long. Let's start with the same
task but using `blastp`. To begin, let's organize our thoughts and files
in the Assignment 2 repository.  __If you have not already done so__, clone the repository now.  You can find the appropriate URL by going to github

	cd 
	git clone [your URL here]
	
	cd [your directory name here]
	ls #to see what files are already in the repo

* Keep a list of your the commands that you used in your lab2_notebook.md file.  This is for your use and will not be graded.
* Answer the questions in the assignment2-worksheet.md file.

Before you can use `blastp` you need to format the BLAST database. A
FASTA file can be turned into a BLAST database with `formatdb`. Try the
following command to see the command options.

	formatdb --help

As it turns out, we only need a couple of these. Let's make an alias to
a FASTA proteome, format it as a BLAST database, and then observe the
new files.

	ln -s ~/Assignment_1/Data/Species/A.thaliana/protein.fa ./plantpep
	formatdb -i plantpep -p T
	
__although it is fine to have these in your directory for this lab DO NOT ADD THEM TO THE REPOSITORY__  That is DO NOT type `git add plantpep` etc.  We do not want to download a bunch of huge files when we grade your assignment.

You also will need to make a `wormpep` database because you'll need to search the worm against the plant and vice-versa. However when you try this with wormpep you will get many error messages.  If you get sick of seeing all the error messages, type control-C. `blastp` expects the FASTA headers to be formatted in a very specific manner. It turns out that almost everyone formats the header differently even though there are certain conventions that should be followed. That's what real-world bioinformatics looks like sometimes. Here's how to quiet those error messages.

	formatdb -i wormpep -p T 2> /dev/null

The `2>` token redirects the stderr stream to a file instead of the
screen. `/dev/null` is the name of a dummy file that is never made. It's
like a black hole for data: stuff goes in but never comes out. Now let's
observe that there are a bunch of new files you just made with the
`formatdb` command.

	ls -lrt

Now let's try aligning our previous favorite protein (`B0213.10`) to the plant proteome. Make an alias here and call it `P450`. We know from Lab 1
that comparing a single C. elegans protein to a entire proteome can take
a couple minutes, and that if you attempted to align two whole
proteomes, it would take weeks. Let's align P450 to the entire D.
melanogaster proteome with BLAST and see how long it takes to run.

    time blastp -db plantpep -query P450 > default.blast

That was fast! `blastp` is clearly much faster than `water`. Not only
that, if you look at the files, you will see that a BLAST report gives
you some statistics about the search. The E-value is sort of similar to
what you would get if you did some shuffling experiments and determined
how different the alignment was from the random background.

There are a lot of parameters that control the speed of BLAST. The most important
of these are the seeding parameters. The default `blastp` search uses a
word_size of 3 and a threshold of 11. Let's experiment with the
threshold value from 10 to 15.

First figure out how to change the threshold by getting help on blastp:

    blastp -help | more

Now, instead of typing the command 6 times, lets use a __for loop__. You may need to review [for loops]({{site.baseurl}}/{% post_url 2018-04-12-0-bash_for_loops %}).  

I have started the code required for the for loop below. You will need to complete it to get this to run.  

	for T in 10 11 12 13 14 15
			echo "Starting blastp with threshold $T"
			time blastp [YOUR CODE HERE]
		end

Look at the files created

    ls -lrt

Now calculate how long it will take to search a proteome against another
proteome using the parameters above and add these to your notebook and make a table to answer the assignment question.

---------------------------------------------------------------------------

When you're searching 1 protein against a database, the human readable
format of BLAST looks good. But if you're going to post-process the file
with Unix tools, you want the file to be easily parseable. So we're
going to change the output format to something different. Go back and do
the previous BLAST search of P450 vs pep, but add the following
command line option `-outfmt 7` (this changes the output format to be
tab-delimited fields with a little header information that tells you
what each column is). Feel free to explore other output formats (because
you're a scientist and scientists are curious creatures and not robots).

Now add another option: `-max_target_seqs 4`. This will make it so that
BLAST only reports the four best alignments.

Time to set up the searches.  Unlike with WATER, BLAST allows us to search once proteome against the other. We're going to switch to output format 6 because we don't really care about the header lines.

	blastp -query wormpep -db plantpep -outfmt 6 -max_target_seqs 4
	
_Of course you should direct the output to a file rather than your screen_

_Be sure to run the search in both directions._

If you're impatient, you can speed up the search by changing
`-threshold` as you did above. But is that a good idea? You decide. You
can also speed up the program by telling it to use more threads with the
`-num_threads` option. If you're curious about these options (and
others), it's best to try them on something smaller than searching one
proteome against the other. You could make miniature versions of each
using `head`, for example.  _Because our virtual machines only have two CPUs I do NOT recommend changing `num-threads` in this case.

While you're waiting for the jobs to complete, you might want to tackle
the 2nd half of this lab farther below.

-------------------------------------------------------------------------

Once the jobs are complete use `less` to examine the outputs of each
(actually, you can do this even before the jobs are complete). Use the
`less` search feature (forward slash key) to find proteins in one file
that match proteins in the other file. Find some putative orthologs.


## Part 2: Paralogs

In this part, we are going to look for the **paralogs** of a couple worm
genes: T21B10.2a and B0213.10. That is, we are going to search the these
proteins against the C. elegans proteome to look for sequences that
arose by duplication within the same genome. A gene may have from 0 to
many paralogs. A simple, but flawed way to determine which alignments
represent paralogs is to use an E-value cutoff. But it's a good place to
start. Create the files as needed given the command lines below and then
do the searches.

	blastp -db wormpep -query T21B10.2a > T21B10.2a.blastp
	blastp -db wormpep -query B0213.10 > B0213.10.blastp

Inspect the BLAST report with `less`. Do all the sequences look like
they are highly related? You may want to change the E-value cutoff to
something higher or lower than 1e-10.

